### PR TITLE
Enforce host memory for DT_INT32 inputs of functional ops

### PR DESCRIPTION
**Pull request by ekuznetsov139**
This PR amends the memory type (host/device) assignment logic.

Prior to this fix, there were several situations where a GPU-allocated DT_INT32 tensor could arrive into a kernel via a HostMemory input (typically when that kernel was inside an If or a StatelessWhile subgraph). This _usually_ worked fine, because it's common for the GPU memory to be mapped into the host memory address space anyway, so the kernel could still access the data (which is why the problem was not noticed for so long). But, on some devices and some systems, memory mapping is partial or unavailable, and, on those, the kernel would crash.

### CODE REVIEW

**Comment by cheshire** [comment]
That's a pretty huge non-ROCm specific change. Could you show the test cases where it is crashing otherwise? I also thing the tag `[ROCm]` is somewhat misleading here.

**Review by cheshire — CHANGES_REQUESTED** [review]
The `ROCm` tag is misleading here. Are there crashing test cases?

**Comment by ekuznetsov139** [comment]
It crashes on some ROCm devices in the convert_to_constants test.

**Comment by cheshire** [comment]
Sure, but you are changing the logic for everything. Do you know why is it crashing only on ROCm?

**Comment by ekuznetsov139** [comment]
It crashes because some ROCm devices do not map all GPU memory into host space. It might be crashing with some NV devices as well but I don't have the hardware to test it.
I don't think the change is all that huge, apart from formatting, it should only affect inputs to functional ops like StatelessWhile.

**Comment by cheshire** [comment]
@sanjoy for comments

**Comment by sanjoy** [comment]
I think @agarwal-ashish is a better reviewer here.

**Comment by gbaned** [comment]
@agarwal-ashish Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@saxenasaurabh Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@ezhulenev  Can you please review this PR ? Thanks!

File: tensorflow/core/framework/memory_types.cc
Comment (alextp) [inline tensorflow/core/framework/memory_types.cc:139]: People reading this comment after the PR is merged will likely have no idea what the old behavior is. Can you change the comment to describe the bhavior.

**Review by alextp — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@ekuznetsov139  Can you please check @alextp's comments and resolve conflicts?. Thanks!

**Comment by gbaned** [comment]
@ekuznetsov139  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@ekuznetsov139  Can you please resolve conflicts? Thanks!

File: tensorflow/core/framework/memory_types.cc
Comment (allenlavoie) [inline tensorflow/core/framework/memory_types.cc:67]: Rather than a list of ops, can we do something like:

```
bool IsFunctionCallOp(const NodeDef& node) {
  for (const auto& attr : node.attr()) {
    if (attr.second.value_case() == AttrValue::kFunc){
      return true;
    }
  }
  return false;
}
```

and then pass a NodeDef here? That should cover everything in the list now.

File: tensorflow/core/framework/memory_types.cc
Comment (allenlavoie) [inline tensorflow/core/framework/memory_types.cc:145]: Is `GetTotal(inp_names)` ever going to be different than `inp_dtypes.size()`? If not can we use the latter all the time?

File: tensorflow/core/framework/memory_types.cc
Comment (allenlavoie) [inline tensorflow/core/framework/memory_types.cc:150]: "can this ever be false?" -> Not in this branch of the if, but kdef will be false for functions that are called by setting the op type to the function name. We mostly use that for calling from eager, but there are graphs/unit tests floating around that use it in graphs too. I think it's safe to assume everything else has a kernel and remove this check.

Higher-level question: given that If/While/PartitionedCall/etc. do have kernels, should we be respecting the kernel annotations rather than special-casing int32 for them in the host_memory_required lambda? Did we just not properly annotate the int32 kernels yet?

File: tensorflow/core/framework/memory_types.cc
Comment (allenlavoie) [inline tensorflow/core/framework/memory_types.cc:53]: Are these vlogs going to make any sense if someone hasn't carefully read the code? Probably want to make them more verbose if we keep them.

**Comment by gbaned** [comment]
@ekuznetsov139  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

File: tensorflow/core/framework/memory_types.cc
Comment (ekuznetsov139) [reply tensorflow/core/framework/memory_types.cc:67]: Possibly. A quick test found a few ops which pass the this test but aren't on the list (MapDataset, FlatMapDataset) but those may need to be added to the list anyway.

File: tensorflow/core/framework/memory_types.cc
Comment (ekuznetsov139) [reply tensorflow/core/framework/memory_types.cc:145]: I am fairly sure that they used to be different in some situations when this PR was initially developed (back in April), in cases of ops with variable numbers of inputs/outputs or with list inputs/outputs. However, I can't reproduce this now. Something may have changed in the framework in in the interim.

File: tensorflow/core/framework/memory_types.cc
Comment (ekuznetsov139) [reply tensorflow/core/framework/memory_types.cc:150]: We need to do both.
Some inputs of If/While/etc. go direct into the corresponding kernel, and we mark them as host by looking at the annotations.
Other are passed through to subgraphs untouched. If I remember correctly (and again, it's been 6 months), what used to happen was that some DT_INT32 tensor was passed to If or While, and then materialized inside a subgraph as an output of an Arg(?) op that connected to a Less op. Less op expected host memory input, but If op was the one that determined tensor's location, and, as originally implemented, it did not mark the tensor as host.

Comment (allenlavoie) [reply tensorflow/core/framework/memory_types.cc:150]: So what we really want to do is inspect the function and find the kernel(s) that eventually consume the tensor, check if they have a host memory annotation, and follow that. Or possibly easier, leave the tensor where it is an lazily copy it if necessary. Assuming that kernels will want int32-dtype tensors on host is pretty sad.

@qqfish we have lazy tensor copy for function calls from eager (IIUC). Do you know if it would be feasible for us to do something similar for tensors passed from graph mode into a functional op's kernel?

Comment (ekuznetsov139) [reply tensorflow/core/framework/memory_types.cc:150]: I am not sure if the information about the ultimate consumer is even available at the time this is executed (the consumer may be in a different graph). 
Keeping the tensor in its default location  is suboptimal, because most consumers of DT_INT32 expect it to be on host (because it is a shape or a single integer.)
The optimal and robust solution would be to put INT32 outputs of functional ops on host, and to add checking when subgraphs are connected, in case it is actually expected on the device.

Comment (allenlavoie) [reply tensorflow/core/framework/memory_types.cc:150]: Not "copy it to the device", but "don't do any copies until we hit a real kernel". Functional operations shouldn't have an opinion about where their inputs are, since kernels are free to consume int32 tensors on-device. And likewise they should leave their outputs where they are.

I assume we get this behavior when we inline control flow operations (which happens almost always for PartitionedCall/While/If). It'd be sad if failure to inline lead to special-casing int32. Just want to assess feasibility.

Comment (qqfish) [reply tensorflow/core/framework/memory_types.cc:150]: Right now, lazy tensor copy in eager only works for cross-host tensor copy. And I don't think we should reuse here because I think we should only use lazy tensor copy in eager mode since in eager mode tensors are copied implicitly for usability. In graph mode, it makes more sense to let the placer to insert Send/Recv ops explicitly since it can make graph debugging easier.

So this change actually make sense to me. If we are going to support more functional op, I think we should consider adding a new kernel/op registration flag instead of hard coding op name here.

Comment (allenlavoie) [reply tensorflow/core/framework/memory_types.cc:150]: The placer could still insert send/recv ops explicitly, it's just a question of whether we insert them before a functional op or in the body of the functional op's function (or both). So probably not the same infrastructure as eager lazy tensor copy, but instead just some bookkeeping to let functional ops know about host/device and let them forward that to their calls to the placer for the functions they're wrapping.

We allow kernels to produce/consume int32 tensors on device, so it'd be nice if feeding an int32 device tensor through a functional op didn't change that.

**Comment by gbaned** [comment]
@ekuznetsov139 Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by ekuznetsov139** [comment]
It seems that we're waiting for further feedback from TF devs.

**Comment by gbaned** [comment]
> It seems that we're waiting for further feedback from TF devs.

@allenlavoie Can you please take a look on above comments from @ekuznetsov139. Thanks!

**Comment by allenlavoie** [comment]
What more did you want feedback on?

**Comment by ekuznetsov139** [comment]
Mainly, do you want me to make any specific changes to this PR, and do you even want to merge it at all (as opposed to having someone at your end rewrite that entire code).

**Comment by allenlavoie** [comment]
Unless there's a pressing need I'd rather we didn't make int32 more special than it already is. I doubt users will run into this since control flow is typically inlined rather than these functional kernels running directly.

So if there's a great need for more int32 special casing please make that case, but otherwise I'd say we should implement a more satisfying solution (figure out where downstream kernels want inputs placed / leave them where they are). Of course you're free to leave the latter to someone else.

**Comment by gbaned** [comment]
@ekuznetsov139  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

**Pull request closed without merging**
